File: tools/sequence_viewer/archive.py
from __future__ import annotations

import csv
import struct
from pathlib import Path
from typing import Optional

from .types import Address


class RelocArchive:
    """Repository over unpacked Sagas reloc files, matching sagas::n64::RelocArchive."""

    def __init__(self, assets: Path) -> None:
        self.assets = Path(assets)
        self._files: dict[int, bytes] = {}
        self._links: dict[tuple[int, int], Address] = {}
        self._links_loaded: set[int] = set()
        self.symbols: dict[str, Address] = {}
        self.file_names: dict[int, str] = {}
        self._load_symbols()
        self._load_manifest()

    def _load_symbols(self) -> None:
        path = self.assets / "reloc" / "symbols.tsv"
        with path.open(newline="", encoding="utf-8") as stream:
            reader = csv.DictReader(stream, delimiter="\t")
            for row in reader:
                self.symbols[row["symbol"]] = Address(int(row["file"]), int(row["offset"]))

    def _load_manifest(self) -> None:
        path = self.assets / "reloc" / "manifest.tsv"
        if not path.exists():
            return
        with path.open(newline="", encoding="utf-8") as stream:
            reader = csv.DictReader(stream, delimiter="\t")
            for row in reader:
                self.file_names[int(row["id"])] = row["name"]

    def symbol(self, name: str) -> Optional[Address]:
        if not name or name == "-":
            return None
        return self.symbols.get(name)

    def require_symbol(self, name: str) -> Address:
        found = self.symbol(name)
        if found is None:
            raise KeyError(f"missing reloc symbol: {name}")
        return found

    def bytes_of(self, file_id: int) -> bytes:
        blob = self._files.get(file_id)
        if blob is not None:
            return blob
        path = self.assets / "reloc" / f"{file_id:04d}.bin"
        blob = path.read_bytes()
        self._files[file_id] = blob
        return blob

    def _load_links(self, file_id: int) -> None:
        if file_id in self._links_loaded:
            return
        self._links_loaded.add(file_id)
        path = self.assets / "reloc" / f"{file_id:04d}.links.tsv"
        if not path.exists():
            return
        with path.open(newline="", encoding="utf-8") as stream:
            reader = csv.DictReader(stream, delimiter="\t")
            for row in reader:
                location = Address(file_id, int(row["location"]))
                self._links[(location.file, location.offset)] = Address(
                    int(row["target_file"]), int(row["target_offset"]))

    def resolve(self, pointer_word: Address) -> Optional[Address]:
        self._load_links(pointer_word.file)
        return self._links.get((pointer_word.file, pointer_word.offset))

    def u32(self, address: Address) -> int:
        data = self.bytes_of(address.file)
        end = address.offset + 4
        if end > len(data):
            raise IndexError(f"N64 u32 read past end of reloc {address}")
        return struct.unpack_from(">I", data, address.offset)[0]

    def s16(self, address: Address) -> int:
        data = self.bytes_of(address.file)
        if address.offset + 2 > len(data):
            raise IndexError(f"N64 s16 read past end of reloc {address}")
        return struct.unpack_from(">h", data, address.offset)[0]

    def u16(self, address: Address) -> int:
        data = self.bytes_of(address.file)
        if address.offset + 2 > len(data):
            raise IndexError(f"N64 u16 read past end of reloc {address}")
        return struct.unpack_from(">H", data, address.offset)[0]

    def f32(self, address: Address) -> float:
        return struct.unpack(">f", struct.pack(">I", self.u32(address)))[0]

    def u8(self, address: Address) -> int:
        data = self.bytes_of(address.file)
        if address.offset >= len(data):
            return 0
        return data[address.offset]

    def file_name(self, file_id: int) -> str:
        return self.file_names.get(file_id, f"reloc{file_id:04d}")

File: src/n64/Archive.cpp
#include <sagas/n64/Archive.hpp>

#include <bit>
#include <fstream>
#include <iomanip>
#include <sstream>
#include <stdexcept>

namespace sagas::n64 {
namespace {

std::string reloc_name(std::uint32_t file, std::string_view suffix) {
    std::ostringstream name;
    name << "reloc/" << std::setw(4) << std::setfill('0') << file << suffix;
    return name.str();
}

std::vector<std::string> fields(const std::string& line) {
    std::vector<std::string> result;
    std::stringstream stream(line);
    std::string value;
    while (std::getline(stream, value, '\t')) result.push_back(value);
    return result;
}

} // namespace

RelocArchive::RelocArchive(AssetRepository& assets) : assets_(assets) {
    std::ifstream input(assets_.path("reloc/symbols.tsv"));
    std::string line;
    std::getline(input, line);
    while (std::getline(input, line)) {
        const auto item = fields(line);
        if (item.size() == 3)
            symbols_.emplace(item[0], Address{static_cast<std::uint32_t>(std::stoul(item[1])),
                                               static_cast<std::uint32_t>(std::stoul(item[2]))});
    }
}

std::optional<Address> RelocArchive::symbol(std::string_view name) const {
    const std::scoped_lock lock(mutex_);
    if (const auto found = symbols_.find(std::string(name)); found != symbols_.end()) return found->second;
    return {};
}

std::span<const std::byte> RelocArchive::bytes(std::uint32_t file) {
    const std::scoped_lock lock(mutex_);
    if (auto found = files_.find(file); found != files_.end()) return *found->second;
    auto blob = assets_.blob(reloc_name(file, ".bin"));
    files_.emplace(file, blob);
    return *blob;
}

void RelocArchive::load_links(std::uint32_t file) {
    if (links_loaded_.contains(file)) return;
    links_loaded_[file] = true;
    std::ifstream input(assets_.path(reloc_name(file, ".links.tsv")));
    std::string line;
    std::getline(input, line);
    while (std::getline(input, line)) {
        const auto item = fields(line);
        if (item.size() != 3) continue;
        const Address location{file, static_cast<std::uint32_t>(std::stoul(item[0]))};
        links_.emplace(key(location), Address{static_cast<std::uint32_t>(std::stoul(item[1])),
                                              static_cast<std::uint32_t>(std::stoul(item[2]))});
    }
}

std::optional<Address> RelocArchive::resolve(Address pointer_word) {
    const std::scoped_lock lock(mutex_);
    load_links(pointer_word.file);
    if (const auto found = links_.find(key(pointer_word)); found != links_.end()) return found->second;
    return {};
}

std::uint32_t RelocArchive::u32(Address address) {
    const std::scoped_lock lock(mutex_);
    const auto data = bytes(address.file);
    if (address.offset + 4 > data.size()) throw std::out_of_range("N64 u32 read at "+std::to_string(address.file)+":"+std::to_string(address.offset)+" (size "+std::to_string(data.size())+")");
    const auto* p = data.data() + address.offset;
    return (std::to_integer<std::uint32_t>(p[0]) << 24) |
           (std::to_integer<std::uint32_t>(p[1]) << 16) |
           (std::to_integer<std::uint32_t>(p[2]) << 8) | std::to_integer<std::uint32_t>(p[3]);
}

std::int16_t RelocArchive::s16(Address address) {
    const std::scoped_lock lock(mutex_);
    const auto data=bytes(address.file);
    if (address.offset+2>data.size()) throw std::out_of_range("N64 s16 read at "+
        std::to_string(address.file)+":"+std::to_string(address.offset));
    return static_cast<std::int16_t>((std::to_integer<unsigned>(data[address.offset])<<8) |
                                     std::to_integer<unsigned>(data[address.offset+1]));
}
float RelocArchive::f32(Address address) {
    const std::scoped_lock lock(mutex_);
    return std::bit_cast<float>(u32(address));
}

} // namespace sagas::n64
